runtime/ts/storage/pouchdb/pouch-db-variable.ts

```diff
@@ -0,0 +1,340 @@
+import {assert} from '../../../../platform/assert-web.js';
+import {PouchDbStorageProvider} from "./pouch-db-storage-provider";
+import {PouchDbStorage} from "./pouch-db-storage";
+
+/**
+ * Loosely defines the object stored.
+ */
+interface VariableStorage {
+  /** The id of this Variable */
+  id: string;
+  /** A reference to another storage key, used for refernce mode */
+  storageKey?: string;
+  /** A the actual value of the data */
+  rawData?: {};
+}
+
+interface VariableStorageMutator {
+  (value: VariableStorage) : VariableStorage;
+}
+
+export class PouchDbVariable extends PouchDbStorageProvider {
+  private _stored: VariableStorage|null = null;
+  private localKeyId = 0;
+  
+  constructor(type, storageEngine: PouchDbStorage, name: string, id: string, key: string) {
+    super(type, storageEngine, name, id, key);
+    this.backingStore = null;
+  }
+
+  backingType() {
+    return this.type;
+  }
+
+  clone() {
+    const variable = new PouchDbVariable(this.type, this.storageEngine, this.name, this.id, null);
+    variable.cloneFrom(this);
+    return variable;
+  }
+
+  async cloneFrom(handle) {
+    this.referenceMode = handle.referenceMode;
+    const literal = await handle.toLiteral();
+
+    if (this.referenceMode && literal.model.length > 0) {
+      // cloneFrom the backing store data by reading the model and writing it out.
+      await Promise.all([this.ensureBackingStore(), handle.ensureBackingStore()]);
+
+      literal.model = literal.model.map(({id, value}) => ({id, value: {id: value.id, storageKey: this.backingStore.storageKey}}));
+      const underlying = await handle.backingStore.getMultiple(literal.model.map(({id}) => id));
+      await this.backingStore.storeMultiple(underlying, [this.storageKey]);
+    }
+    
+    await this.fromLiteral(literal);
+
+    // TODO(lindner): ask shane why this doesn't fire 'change' events like firebase does...
```

Comment:
Maybe it should :)

Better testing post-M1 will tell us.

Reply:
Ack.